_drafts/learn-doobie-for-great-good.md

```diff
@@ -0,0 +1,961 @@
+
+```scala
+def findAllDirectorsProgram(): IO[List[Director]] = {
+  val findAllDirectors: fs2.Stream[doobie.ConnectionIO, Director] =
+    sql"select id, name, last_name from directors".query[Director].stream
+  findAllDirectors.compile.toList.transact(xa)
+}
+```
+
+However, as we may expect, running the above program generates an error since the compiler cannot find any type class instance for `Read[Director]`. So, let's define the missing type class instances:
+
+```scala
+object Director {
+  implicit val directorRead: Read[Director] =
+    Read[(Int, String, String)].map { case (id, name, lastname) =>
+      Director(DirectorId(id), DirectorName(name), DirectorLastName(lastname))
+    }
+  implicit val directorWrite: Write[Director] =
+    Write[(Int, String, String)].contramap { director =>
+      (director.id.id, director.name.name, director.lastName.lastName)
+    }
+}
+```
+
+_Et voilà_, now the program runs without any error.
+
+## 7. Handling Joins
+
+Until now, we presented some very straightforward examples of queries. However, we can also handle joins. The good news is that Doobie takes join between tables in a very natural way.
+
+Let's say we want to find a movie by its name. We want to retrieve also the director's information and the list of actors that played in the film. Using the ER model we presented at the beginning of the article, we have to join three tables: the `movies` table, the `directors` table, and the `actors` table. Here is how we can implement it in Doobie:
+
+```scala
+def findMovieByNameProgram(movieName: String): IO[Option[Movie]] = {
+  val query = sql"""
+       |SELECT m.id,
+       |       m.title,
+       |       m.year_of_production,
+       |       array_agg(a.name) as actors,
+       |       d.name
+       |FROM movies m
+       |JOIN movies_actors ma ON m.id = ma.movie_id
+       |JOIN actors a ON ma.actor_id = a.id
+       |JOIN directors d ON m.director_id = d.id
+       |WHERE m.title = $movieName
+       |GROUP BY (m.id,
+       |          m.title,
+       |          m.year_of_production,
+       |          d.name,
+       |          d.last_name)
+       |""".stripMargin
+    .query[Movie]
+    .option
+  query.transact(xa)
+}
+```
+
+Since the `actor` table extracts many rows for every movie, we used a `GROUP BY` operation and the `array_agg` Postgres function, which creates an array from the actors' names.
+
+However, the array type is not SQL standard. So, to let Doobie map the array type to a Scala `List`, we need to import the `doobie.postgres._` and `doobie.postgres.implicits._` packages, belonging to the `doobie-postgres` library.
+
+As we said, the array type is not standard, and it could happen to implement the above program in a database that doesn't support arrays. In this case, the only solution left is to **perform the join manually, which means splitting the original query into three different queries and joining the data programmatically**:
```

Comment:
it could happen to implement the above program in a database that doesn't support arrays 
->
the database we're using might not support arrays